### ISSUE TITLE
Hung orchestration after child function completes

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#518]
I came across hung orchestration issues when testing on consumption and found this issue (#460), so I did some testing on a single B2 VM. I'm not sure if it's related to this, but I have an eternal orchestration that cycles through 1-2k items, at 150 per orchestration, then uses continue as new until the 1-2k are all processed.

I haven't been able to get through the entire orchestration yet. I've tested this probably 8-10 times now, and each time, eventually the orchestration gets "stuck". There is also no pattern to it. The last trace is always one of the activities executed, and then nothing. It's not always the same activity, and there aren't any exceptions thrown. 

If I restart the functions app, it will begin running again, but if I don't restart it, it will be stuck right at that last executed activity.

Any ideas?

**EDIT**: More tests this morning. It looks more like the issue is that a sub orchestration is failing with an exception, but the parent orchestration stays hung in running status, and no exceptions are traced, so it appears like the overall orchestration is in a "stuck"(running) state instead of having a cascading fail. I found this by looking in the DurableFunctionsHubInstances table, but there is nothing in application insights to indicate the failure.  FWIW, The sub orchestrations are being thrown into a list of tasks and then await Task.WhenAll, which doesn't seem to get hit as a result of the aforementioned exception within one of those tasks. 

When the functions app is restarted, it recognizes the failure and correctly updates the parent orchestration from running to failed, but without the restart, it will stay in "running" state. Another thing to note is that on restart, the failures are just informational and not logged as error, which I saw brought up in another issue here.

**EDIT 2** (sorry): Just had it happen again - stuck on a sub orchestration just hanging. This time there was no exception or failure though. It was just stuck. Last trace was "executed" of an activity within that sub orchestration. Restarting the app service found it start with executing->completed->executed of said sub orchestration, which might lead to the conclusion that after the last activity of the sub orchestration completed, the sub orchestration failed to return to the parent orchestration until the app service was restarted. Once restarted, the parent activity finished executing the hung sub orchestration and continued on as expected. 

I didn't have these issues until I recently refactored multiple activities into sub orchestrations. When there was only the parent orchestration with everything else in activities it worked fine.

**LAST EDIT**: The parent activity from my second edit completed, and the last thing in trace is:

`Function 'ParentOrchestration (Orchestrator)' completed. ContinuedAsNew: True. IsReplay: False. Output: (null). State: Completed. HubName: DurableFunctionsHub. `

But it has not continued. It is hanging once again. In the DurableFunctionsHubInstances table, it has a status of running. I gave it 30 minutes, now I just restarted the functions app. Upon reboot, it replayed the last half dozen traces(with IsReply:false) and continued on as it should have in the first place and is now running. Not sure what is happening in all these instances, but it seems restarting the app corrects itself each time.

_Originally posted by @VividDigital in https://github.com/Azure/azure-functions-durable-extension/issues/460#issuecomment-435906135_

**Comment by cgillum** [comment]
@VividDigital This sounds a lot like this issue: https://github.com/Azure/azure-functions-durable-extension/issues/502. Can you take a look and see if it sounds related?  If so, we have a fix coming in the next release, which should be out in a week or two.

**Comment by VividDigital** [comment]
Yes I think it could have something to do with that as well since it makes the most sense. Also, I made this [stack overflow](https://stackoverflow.com/questions/53197710/durable-functions-sub-orchestration-never-returns-to-parent-orchestration) question as well which may be somewhat related, and I really think it has to do with the message size or passing of the message in some way. I can replicate it almost every time now, and if I change the amount of items I'm sending back to 50 instead of 150, the sub orchestration does not hang.

**Comment by VividDigital** [comment]
Thanks Chris. One other piece of info that may be relevant is in these
situations, the live metrics on the app had no activity, but always had a
large amount of committed memory (~2GB). This was consistent each time.

Ryan Payne
416.904.7238
http://thevividweb.com

**Comment by jcmartinez23** [comment]
We have exactly the same problem, only it's not with eternal orchestrations. We have a large amount of suborchestrations fired that handle files of about 50MB and it gets stuck constantly. The behaviour is as described by @VividDigital, the app shows no activity or exceptions but memory consumption it's pretty high.

If I look into the workitems queue, I can see that the messages are picked up by the runtime, but nothing is being executed on the workers. One thing that may help: I started looking through the logs on the file share and saw some exceptions that I didn't see in AI. There was one entry that stood out:

```
2018-11-14T17:02:21.865 [Trace] BlackList CSV item with CorrelationId=dcda7e30-d5ef-48d7-93b4-3d892238da7d successfully parsed to Object
2018-11-14T17:02:21.865 [Trace] BlackList object with CorrelationId=dcda7e30-d5ef-48d7-93b4-3d892238da7d successfully parsed to JSON
2018-11-14T17:02:23.232 [Trace] BlackList CSV item with CorrelationId=dcda7e30-d5ef-48d7-93b4-3d892238da7d successfully parsed to Object
2018-11-14T17:02:23.232 [Trace] BlackList object with CorrelationId=dcda7e30-d5ef-48d7-93b4-3d892238da7d successfully parsed to JSON
2018-11-14T17:02:43.069 [Error] Executed 'ConvertCsvToJsonActivity' (Failed, Id=419f1280-d9bb-4c76-8650-068e728d1ef9)
Exception of type 'System.OutOfMemoryException' was thrown.
2018-11-14T17:02:52.823 [Debug] Workers Directory set to: D:\Program Files (x86)\SiteExtensions\Functions\2.0.12175\32bit\workers
2018-11-14T17:02:52.834 [Debug] Found worker config: D:\Program Files (x86)\SiteExtensions\Functions\2.0.12175\32bit\workers\java\worker.config.json
2018-11-14T17:02:52.852 [Debug] Will load worker provider for language: java
2018-11-14T17:02:52.852 [Debug] Found worker config: D:\Program Files (x86)\SiteExtensions\Functions\2.0.12175\32bit\workers\node\worker.config.json
2018-11-14T17:02:52.858 [Debug] Will load worker provider for language: node
2018-11-14T17:02:52.859 [Debug] Worker path for language worker java: D:\Program Files (x86)\SiteExtensions\Functions\2.0.12175\32bit\workers\java
2018-11-14T17:02:52.859 [Debug] Worker path for language worker node: D:\Program Files (x86)\SiteExtensions\Functions\2.0.12175\32bit\workers\node
2018-11-14T17:02:53.609 [Information] Initializing extension with the following settings: Initializing extension with the following settings:
```

As you can see, a 'System.OutOfMemoryException' is thrown and then everything restarts. The problem is that on local executions this doesn't happen and that I only can see this Exception in these log files without any kind of stack trace.

We are getting this behaviour on a App Service Plan with a fixed VM count of 4.
